### ISSUE TITLE
Categorical Reparameterization with GUMBEL-SOFTMAX

### ISSUE THREAD

**Issue by patrick-g-zhang** [patrick-g-zhang/papers-i-read#7]
## Motivation
- stochastic neural network rarely use categorical latent variables due to the inability to back propagate through samples
- Gumbel-Softmax distribution make it work
### Path derivative gradient estimators aka re-parameterized trick
if z is not reparamterizable, how to
- ST (high variances)
- mixture model and marginalize the binary variables

### Score Function-Based Gradient Estimator
- does not require back propagating through f or the sample z
- high variance

### SEMI-SUPERVISED GENERATIVE MODELS

![img](https://pbs.twimg.com/media/Ee5da8DUYAAgMq3?format=jpg&name=large)
### Gumbel-Softmax Estimator
- replacing non-differentiable categorical samples with a differentiable approximation 
- there is a trade off between small temperatures, where samples are close to one-hot but the variance of the gradients is large
-  start at a high temperature and anneal to a small but non-zero temperature
![img](https://pbs.twimg.com/media/Ed0vhGiUwAASMNX?format=png&name=large)

### Straight-Through Gumbel-Softmax Estimator
- we discretize y using arg max but use our continuous approximation in the backward pass by approximating \Deltaθz ≈ \Deltaθy. 
- biased path derivative estimator
- samples sparse even when the temperature is high

## Experiments
- compare different gradients estimators:

1. Gumbel-Softmax
2. ST Gumbel-Softmax
3. Score Function
4. ...

- two tasks
1.  structured output prediction and
2. variational training of generative models. 

### STRUCTURED OUTPUT PREDICTION WITH STOCHASTIC BINARY NETWORKS
The objective of structured output prediction is to predict the lower half of a 28 × 28 MNIST digit **given the top half of the image** (14×28).    
![img](https://pbs.twimg.com/media/Ee7Rc9WUwAI2F9E?format=jpg&name=large)

### GENERATIVE MODELING WITH VARIATIONAL AUTOENCODERS
- pure variational autoencoder
- model the latent variable as a single hidden layer with 20 Bernoulli variables or 20 categorical variables
![img](https://pbs.twimg.com/media/Ee7SzgZUcAc4yAd?format=jpg&name=small)

### GENERATIVE SEMI-SUPERVISED CLASSIFICATION
- compare original marginalization-based inference approach and single-sample inference with Gumbel-Softmax and ST Gumbel-Softmax
- 100 labeled examples and 50000 unlabeled examples
- result shows they performed equally

## Conclusion and My Thoughts
- 提供了一种方法可以让离散随机变量利用reparameterized trick
- low variances of gradients
